core/domain/classifier_services.py

```diff
@@ -576,6 +586,11 @@ def migrate_state_training_jobs(
         exp_id, version=exp_version)
     interaction_id = exploration.states[state_name].interaction.id
 
+    if interaction_id is None:
```

Comment:
Per https://github.com/oppia/oppia/pull/15693#discussion_r939060262 (trying to avoid this discussion being buried in GitHub):

>> The assertion on its own is not sufficient here because someone may write another function that calls this one in the future.
>
> If someone calls a function with Optional type and we have assertion written in the called function, then assertion will caught this situation and throw an assertion error that value cannot be None

But that error will happen at runtime and possibly in production, whereas mypy checks are supposed to be compile-time and developer-time checks. So, it would be good to try and catch as many errors as possible using mypy rather than rely on assertions.

Reply:
> But that error will happen at runtime and possibly in production, whereas mypy checks are supposed to be compile-time and developer-time checks.

Ok, now I have a clear picture of why assert is not preferred. Thanks!!